### PR TITLE
Naming Customization and Cleanup

### PR DESCRIPTION
**Pull request by erikerikson**
## What did you implement:

**Closing:** #2022 

## How did you implement it:

I added `~/lib/plugins/aws/lib/naming.js` as a central definition of naming standards.   All code partaking in naming was altered to use the centralized naming capability (or at least this was the intention).  The notable exceptions to this is the not-provider-specific function name (an overridable default) and the packageService defined names of zip files.
## How can we verify it:

Tests were added for the new code.  That said, the change was intended to be a no-op from a correctness standpoint.  Where names were hard coded, the naming class is now used.  Otherwise, nothing of semantic importance should be changed.  I found it useful to corrupt the centrally defined names in order to identify areas that may have been missed.  The relationship between hardcoded .json files and the centralized naming is the only source of errors that I found.
## Todos:
- [x] Write tests
- [x] Write documentation (N/A)
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [] Change ready for review message below

**_Is this ready for review?:**_ NO

### CODE REVIEW

**Comment by erikerikson** [comment]
This should be ready for review.

Note that I only attempted to centralize naming.  I did not yet attempt to add controls over naming conventions or anything of the like (although these seem vitally important).

I believe that some of the idioms will be unpleasant and should be fixed (particularly the `naming.configure(serverless)` idiom) but I thought it would be better to open those to discussion.

**Comment by erikerikson** [comment]
Reduced coverage is as a result of reduced lines of code.

**Comment by eahefnawy** [comment]
Beautiful! Thanks for all the heavy lifting @erikerikson 😊 ... our codebase was starting to get bloated because of all that naming issues.

Currently reviewing...

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js [hunk 1]
```diff
@@ -1,7 +1,10 @@
 'use strict';
 
-const BbPromise = require('bluebird');
 const _ = require('lodash');
+const BbPromise = require('bluebird');
+const pathLib = require('path');
+
+const naming = require(pathLib.join(__dirname, '..', '..', '..', '..', '..', 'lib', 'naming.js'));
```
Comment (dougmoscrop) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: Could you mix-in naming in to the apiGateway/index.js file to avoid dot dot hell?

Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: Absolutely, I thought I'd removed all instances of this from the code.  Thanks for noticing it.

Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: Oh...  I misread this initially.  I've removed the painful use of path.  There was a set of cases where my search had failed to catch some occurrences.

Reading you as requesting that naming.js be moved into `~/lib/plugins/aws/deploy/compile/events/apiGateway/index.js`, I have to note that the naming module is defined at `~/lib/plugins/aws/lib/naming.js` because the naming conventions need to be applied through the provider specific plugin.  Examples include:
`~/lib/plugins/aws/index.js`
`~/lib/plugins/aws/info/index.js`
`~/lib/plugins/aws/logs/index.js`
`~/lib/plugins/aws/remove/index.js`

**Comment by dougmoscrop** [comment]
Ya I was wondering maybe putting it on provider? Like

serverless.provider.nameXYZ(x)

Make the aws plugin /lib object.assign it to provider or something.

**Comment by erikerikson** [comment]
Okay, I see what you're suggesting.  I'll check but there shouldn't be any trouble doing that.

File: lib/plugins/aws/lib/naming.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+ * also be centralized so that they co-occur and are more correlated spatially within the code base.
+ * That is to say that they are easier to change together.
+ */
+class Naming {
+  // TODO this configure pattern is fugly.  it serves the need to have static and also configured
+  // TODO use patterns without complicating the life of consumers.  In standard (non-test) usage
+  // TODO the frequency of using this method should be low so maybe it doesn't matter.
+  // TODO better suggestions highly welcome
+  configure(serverless) {
+    this.serverless = serverless;
```
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: why not in the constructor like we're doing in all of our plugins?

File: lib/plugins/aws/lib/naming.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  getLogicalRoleName() {
+    return 'IamRoleLambdaExecution';
+  }
+  getPolicyName() {
+    return { // TODO should probably have name ordered and altered as above - see AWS docs
+      'Fn::Join': [
+        '-',
+        [
+          this.serverless.config.stage,
+          // TODO is it a bug here in that the role is not specific to the region?
```
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: I don't think most AWS resource need to be specific to a region, only S3 as far as I know because they're global. But so far we haven't had an issu with that. Other examples are the Lambda name and stack name, it's not specific to the region either.

File: lib/plugins/aws/lib/naming.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+  /*
+   * Lambda Function Naming
+   */
+  getNormalizedLambdaName(functionName) {
+    return this.normalizeName(functionName);
+  }
+  extractLambdaNameFromArn(functionArn) {
+    const splitArn = functionArn.split(':');
+    // TODO the following two lines assumes default function naming?  Is there a better way?
```
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: I think it depends on the context, but this insures we get the actual **Lambda Name** whether the user provided a custom one via the `name` property, or the default name.

File: lib/plugins/aws/lib/naming.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+  }
+  getLogicalApiGatewayApiKeyRegex() {
+    return /^ApiGatewayApiKey/;
+  }
+
+  /*
+   * S3 Bucket Naming
+   */
+  getLogicalDeploymentBucketName() {
+    return 'ServerlessDeploymentBucket';
```
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: would that work with the recent custom bucket feature we merged few days ago? we might need another method or something and this one could be called `getLogicalDeploymentBucketName`

**Comment by erikerikson** [comment]
@dougmoscrop, et al.  I have moved the naming object into an attribute of the SDK object.  This let me get rid of the horrible configure idiom and cleaned things up substantially.  Thanks for the great advice!

File: lib/plugins/aws/lib/naming.js (hunk 2 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: Yes, moved it to the constructor in the latest commit in response to @dougmoscrop's great feedback.  Still absorbing idioms here.

File: lib/plugins/aws/lib/naming.js (hunk 3 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: What was in mind when I wrote that comment is that if you are doing cross-region deployments that you are likely going to need to do cross-region replication or at least backup and that you would generally want to manage that behind region-specific resources.  I suppose so long as we're only talking about default privileges that this is over-thinking the matter.

File: lib/plugins/aws/lib/naming.js (hunk 4 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: if the user's custom name is something like `my-name-with-dashes`, wouldn't it provide the value `dashes`?  We definitely have names with components separated by dashes, which is why it occurred to me.

File: lib/plugins/aws/lib/naming.js (hunk 5 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: Hmm...  I'm not sure.  I have a few commits that need to be merged in.  If it is already merged and there is a test that evaluates custom bucket declaration, then it does work.  If there isn't, then I will run into this when I do the merge.  I've mostly attempted to not change the consequences of the code at all while centralizing the naming so it should have the exact behavior it did before my changes.

**Comment by erikerikson** [comment]
I have merged the latest changes off of master.

@eahefnawy I feel reasonably confident about these changes since they aren't changing any functionality but only placing hard-coded names and name coercion into a centralized module.  This should, of course, be a no-op.

To verify these changes, I have found that adding a prefix and or suffix for all the names in the following locations can turn up a lot of missed corners:
```
~/lib/plugins/aws/lib/naming.js
~/lib/plugins/aws/deploy/lib/core-cloudformation-template.json
~/lib/plugins/aws/deploy/lib/iam-policy-lambda-execution-template.json
~/lib/plugins/aws/deploy/lib/iam-role-lambda-execution-template.json
```
I am getting 100% success on the unit tests which include the custom bucket naming tests as well.

File: lib/plugins/aws/lib/naming.js (hunk 5 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: To explain why: basically, this "LogicalDeploymentBucketName" is 1) the name to call it by in the CFT & 2) overwritten by the logic that identifies that a custom bucket name has been supplied.

I will try and take a look if there's a case there being overlooked in the morning.

File: lib/plugins/aws/lib/naming.js (hunk 3 above)
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: @erikerikson that's a super interesting user case! I haven't thought about that before! Do you think this would cause an issue for you guys at Nordstrom?

File: lib/plugins/aws/lib/naming.js (hunk 5 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: The logical names, by which I mean the attribute names under "Resources" in the CFT, are not affected by the custom bucket changes (see [configureStack.js#L98](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/configureStack.js#L98)).  If a custom bucket name is not supplied then it will use the default value defined in the core template (see [core-cloudformation-template.json](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/core-cloudformation-template.json#L12)).

**Comment by erikerikson** [comment]
Debugging through this with live deployment now.  The region and stage are undefined and causing an issue.

**Comment by erikerikson** [comment]
Turns out that the provider is not propagating the serverless.yml declared stage and/or region to the serverless.config...  Does it seem more appropriate to transfer those on load or to write contingent use code throughout the aws plugin?  The former seems preferable from a terseness standpoint, with a rule that declares "upon loading, the aws plugin will transfer default values declared in the provider specific manner into the global SLS configuration":
```
    // Transfer Provider supplied default values into the global Serverless configuration hive
    if (this.serverless.service.provider.stage) {
      this.serverless.config.stage = this.serverless.config.stage ||
                                      this.serverless.service.provider.stage;
    }
    if (this.serverless.service.provider.region) {
      this.serverless.config.region = this.serverless.config.region ||
                                      this.serverless.service.provider.region;
    }
```

**Comment by erikerikson** [comment]
I successfully deployed using this latest commit but only the hello world default.  I'm operational now in an educational account but don't have anything complicated to real-world test with quite yet.

File: lib/plugins/aws/lib/naming.js (hunk 3 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: In the short term cross-region deployments are not wide spread but we there are definitely those doing this.

File: lib/plugins/aws/lib/naming.js (hunk 4 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: Sorry, I think I wasn't quite clear.  This is from [authorizers.js#30](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/events/apiGateway/lib/authorizers.js#L30).  It is extracting a function name from a given ARN that should identify a lambda that is to be used as the authorizer for apig.  I don't believe that should contain the logical name that was used in the CFT to create the Lambda.  Instead, I believe that if I set a function's custom `name` to be something like `a-name-with-dashes` that I'll end up with an ARN that is something like `arn:aws:lambda:region:${accountId}:function:a-name-with-dashes`.

I would expect this code to return `dashes` rather than `a-name-with-dashes` which I believe is the expected extraction.

Another thought is that the following would seem to be more a appropriate authorizer ARN specification (just adding stage):
`arn:aws:lambda:region:${accountId}:function:a-name-with-dashes:${stage}`
In that case, the stage would be what gets split.

Sorry for being a bit confused and context-poor here, just asking questions from my curiosity while trying to make sure and capture all the name provision and manipulation logic.

**Comment by dougmoscrop** [comment]
Erik I don't know if my refactor PR would make this easier on you? I mean
over everything to json and reduced a lot of duplicate code. I don't mind
rebasing in your changes but I would be doing them by hand. I guess you
would have to do the same. Were both hittinf similar pain points

**Comment by erikerikson** [comment]
Heyya @dougmoscrop, sorry but I hadn't been aware of your PR.  I'll have to take a look.  Broadly speaking, moving out of static text to objects seems positive to me.

I've definitely been attempting not to boil the ocean here, staying narrow and focussed on the naming concern.

Anyway, I'll take a look at what your changes and give this some thought.

File: lib/plugins/aws/lib/naming.js [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  // Probably belongs in a non-aws-specific code asset
+  getServiceEndpointRegex() {
+    return /^ServiceEndpoint/;
+  }
+
+  /*
+   * Stack Naming
+   */
+  getStackName() {
+    return `${this.serverless.service.service}-${this.serverless.config.stage}`;
```
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: @eahefnawy @flomotlik @pmuens - sorry to call out.

`this.serverless.config.stage` appears to not be the natural and expected source of stage and region (referenced elsewhere).  These are elsewhere commonly obtained from the options hive, except that they can come from the provider as well.  Is there a single authoritative source for the region and stage values to use?  Is there a reason that the "right answer" isn't being collected and managed with the serverless.config hive?  Thanks!

File: lib/plugins/aws/lib/naming.js (hunk 4 above)
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: aaaaah yeaah I think you're right 👍 ... in the LOC you provided we're assuming that splitting the dashes will **always** split into only function/stage, but if the function name has dashes the function name itself would be split too.

That's what you mean right? 😄 

does the `extractLambdaNameFromArn` method you provided fixes that?

File: lib/plugins/aws/lib/naming.js (hunk 6 above)
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: hmmm this `this.serverless.config.stage` is a little weird, I dont remember seeing it before. Maybe it slipped in. But you're right, the general source is `this.options.stage` as seen [here in the aws/validate file](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/lib/validate.js#L12), which is common to all aws plugins.

Maybe it's just a matter of naming, and we shouldn't name it `this.options.stage` since it also takes from the provider, but I'm not sure about keeping it in the `serverless.config` object because stage/region are provider specific, while `serverless.config` is more of serverless/service wide config.

**Comment by eahefnawy** [comment]
@erikerikson I've been testing this PR extensively with the following config trying to cover all supported CF resources/logical ids:

```yml
service: serverless-naming

provider:
  stage: beta
  region: us-west-2
  name: aws
  runtime: nodejs4.3
  apiKeys:
    - keyOne
    - keyTwo

functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: users/create
          method: get
          authorizer: auth
      - s3: naming-bucket
      - schedule: rate(10 minutes)
      - sns: naming-topic
  auth:
    handler: handler.auth
```

The PR is pretty solid for all resources, **however** I can't seem to deploy to a stage passed via the CLI. It seems that (as discussed above), we're **always** taking the stage from the provider and not touching the stage passed as an option. The same probably goes for region as well.

Is there a particular reason why you're storing stage/region in `serverless.config`? It's meant more for a serverless/service level config, rather than provider level config. But if you have to, it needs to be also aware of stages/regions passed as options.

File: lib/plugins/aws/lib/naming.js (hunk 6 above)
Comment (flomotlik) [inline lib/plugins/aws/lib/naming.js]: for now `this.options.stage` can be used, but we need to clean this up because its super unintuitive at the moment. Probably a helper function in AWS implementation that reads the current stage from config file, options or other places that it is defined. Similar to the credentials PR @erikerikson recently did

**Comment by erikerikson** [comment]
@eahefnawy thank you for taking the time to do this investigation.  I was thinking that it might help us all if there was a set of integration tests that I could (maybe optionally) run (in addition to the unit tests) which would run through these scenarios.  I feel bad that my efforts have spent your and other member's time.
As discussed in the code comment where I talk about centralized config (made shortly ago), the CLI options are not available in the context of the AWS provider plugin (`class SDK`).  I think the discussion around that could probably happen most effectively there.  Without having a complete understanding to the reasoning for the compartmentalization of configuration that I am observing I would end up taking a randomish stab in the dark.

File: lib/plugins/aws/lib/naming.js [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+class Naming {
+  constructor(serverless) {
+    this.serverless = serverless;
+  }
+
+  /*
+   * General Name Normalization
+   */
+  normalizeName(name) {
+    return `${_.upperFirst(name)}`;
```
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: should this be `_.capitalize` instead?  (this would lower-case every character but the first)  This would be a breaking change since it could change Resource names.

**Comment by eahefnawy** [comment]
@erikerikson aah believe me, reviewing your PRs is the best use of my time! I'm learning a lot 😉

For integration tests, we have the [integration tests suite](https://github.com/serverless/integration-test-suite), but it's sloooooow. I generally like to use the `serverless.yml` template above whenever I'm testing because it's pretty inclusive, but still edge cases has to be well thought through (its an art! 😄  )

File: lib/plugins/aws/lib/naming.js (hunk 6 above)
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: @erikerikson about the stage/region not being available in the context, maybe in that case we could just pass the stage/region to whatever method needs them? this would make those methods "pure" in a sense. That would be similar to how we're doing it with `sdk.request` for example, you just pass it a stage/region. Do you think this setup would fix this current issue?

as for multiple providers, imo multi provider support in itself is not challenging...what's really challenging is **designing a uniform interface for all providers**, simply because each provider is completely different. I honestly am not completely sure how we're going to accomplish that since lots of other providers are still immature in their interface. But we're keeping our eyes open for any opportunities there 😊

File: lib/plugins/aws/lib/naming.js (hunk 7 above)
Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: I'd vote for `_.upperFirst` to leave room for us and the users to use camelCasing

File: lib/plugins/aws/lib/naming.js (hunk 4 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: That is a much clearer statement of what I meant.  😄 
No, I haven't tried to change any of this logic, I am merely working to centralize it.
My thinking is that once this is centralized, we can deep dive on this type of matter as well as enabling configuration and/or plug-ability

File: lib/plugins/aws/lib/naming.js (hunk 7 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: cool, I'll leave this as is

File: lib/plugins/aws/lib/naming.js (hunk 6 above)
Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: re: "maybe in that case we could just pass the stage/region to whatever method needs them?"
Your proposal is what I was attempting to suggest is problematic.  I could do what you suggest and it **would** fix the **immediate** issue.  It creates a long term problem however.  Doing this constrains what can be used as source material for generating a name (admittedly not an immediate concern).  What about the useful `-foo` switch that will be added to the `deploy` command?  Do we add `foo` to this method signature too?  More plausibly, what about the next provider (beyond AWS) that doesn't have a region and a stage parameter?  Do we add that provider's special CLI configurable values to the signature too?  Over the long term, this approach seems to paint us into a corner.

If I'm missing a big point, I apologize, I'm not sure I quite understand what you mean by 'make those methods "pure" in a sense'.  It sounds as if you're suggesting that the interface be made explicit and thereby that it would be very clear how configuration flows and that would provide a kind of purity.  That said, I don't think that the scope of what will be passed is sufficiently compact or well enough known at this point.

What I am proposing is that centrally SLS will have a scheme for managing the configuration.  It already does this with the `Config` object.  That object could be expanded to provide a richer mechanism of interaction.  The interface of the config object can be made explicit and encapsulated so that data protection occurs and each provider cannot mess with shared state (as others see it) except through well defined interfaces.  Each provider will have effectively unpredictable needs.  As a result, we can centrally maintain the configuration with every provider having an altered version of the centralized configuration available only to them.  I'll get back to this and be concrete.

> as for multiple providers

I am not suggesting that we should provide an abstraction that makes interactions with every provider the same (although similarity would be a little helpful from a cognitive load standpoint).  That leads us toward the least common denominator and none of the deeper value each will have over the others.  Instead, I would suggest we can create a means of separating their perspectives on what the configuration is in isolated contexts that would interact (if ever) through well defined data flows.  Something like:

```
class Config { // the central hive
  constructor(id); // this ID would give you your own version of the configuration
  get(key); // this transparently combines the SLS specified configuration state with the Provider specified configuration state
  set(key, value); // this transparently isolates any value changes by a Provider to the provider
}
```

If it became necessary, you could add more:
```
class Config {
  ...
  getFor(id, key);
  setFor(id, key, value);
}
```
And these would alter the value associated with the given key

```
serverless = new Serverless(); // calls serverless.config.set('foo', 'bar'); and serverless.config.set('my', 'val');
aws = new SDK(serverless);
aws.setFoo('caz') // calls serverless.config.set('foo', 'caz'); // serverless.config now contains global = { foo: 'bar' } and aws = { foo: 'caz' }
serverless.printFoo(); // console.log(this.config.get('foo'); // prints "bar"
aws.printFoo(); // console.log(serverless.config.get('foo'); // prints "caz"
serverless.printMy(); // console.log(this.config.get('my'); // prints "val"
aws.printMy(); // console.log(serverless.config.get('my'); // also prints "val"
```

Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: As for what I meant by multi-provider service, I mean that one could do something like:
```
service: myService

provider:
  stage: dev // applies to all providers
  aws:
    runtime: nodejs4.3
    profile: myProfile
    region: us-east-1
  google:
    profile: myGoogleProfile
    region: googleRegion
functions:
  hello:
    handler: handler.hello // uses my google machine learning resource
resources:
  aws:
    Resources: ...
  google:
    // my machine learning resource
```
I would expect that both provider plugins would be loaded and that the google resources would be deployed using the google provider and the aws resources would deployed by that provider.

Comment (eahefnawy) [inline lib/plugins/aws/lib/naming.js]: > What about the useful -foo switch that will be added to the deploy command? Do we add foo to this method signature too?

You're right, and this is why we have the options object available for all plugins. What we wanted to do at first is to init this provider class just like other core classes, but we decided to keep all the provider specific logic away from the core.

while I totally agree with you, imo I think we should just pass as args for now since changing that would require some more discussion and internal effort.

Would love to get @flomotlik input though 😊

Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: Definitely the immediate and lowest change choice is definitely passing options into the provider.

For the long term that means spreading the knowledge across the codebase of how to correctly prioritize and select the various configuration sources.

Comment (erikerikson) [inline lib/plugins/aws/lib/naming.js]: Until the point that centralization occurs, of course.

**Comment by erikerikson** [comment]
Sweet!  Thanks for the integration tests pointer @eahefnawy!  Very glad you're getting value and hope that never changes.

**Comment by erikerikson** [comment]
@eahefnawy The CLI options should now be prefered over configuration in Serverless class or provider specified defaults (via https://github.com/serverless/serverless/pull/2257/commits/ae3b9b746deef3cf62a1f9ffcf1708f80f26ad2f).  Moved the options into the SDK as discussed and cleaned up a bit as appropriate.
Also, merged to get up to date.

I have not run the integration tests as I have to run to a family obligation for the weekend.

**Comment by eahefnawy** [comment]
@erikerikson that's great! thanks tons! 😊 ... we're currently in the process of improving and moving the integration tests to master, so that should make things easier. I'll be playing around with your recent changes again, and run the integration tests against them once we have them in place.

have fun with your family, and thanks for everything 😊 ... will get back to you with updates 👼

**Comment by erikerikson** [comment]
This is merged back to conflict free state but I still have to test the corrupted names configuration (altering the naming class to make sure there's not an accidentally correct name)

File: lib/plugins/aws/tests/naming.js [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+    };
+    serverless = new Serverless(options);
+    sdk = new SDK(serverless, options);
+  });
+
+  it('#normalizeName()', () => {
+    it('should capitalize the first letter', () => {
+      expect(sdk.naming.normalizeName('name')).to.equal('Name');
+    });
+    it('should have no effect on caps', () => {
```
Comment (pmuens) [inline lib/plugins/aws/tests/naming.js]: Maybe some space here so that all tests follow the same pattern. (This pattern might be found in other tests as well)

File: lib/plugins/aws/tests/naming.js [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+    it('should match a name with the prefix', () => {
+      expect(sdk.naming.getServiceEndpointRegex().match('ServiceEndpointForAService')).to.be(true);
+    });
+  });
+  it('#getStackName()', () => {
+    it('should use the service name and stage from the service and config', () => {
+      serverless.service.service = 'myService';
+      expect(sdk.naming.getStackName()).to.equal(`${serverless.service.service}-${options.stage}`);
+    });
+    it('should handle undefined service name', () => {
```
Comment (pmuens) [inline lib/plugins/aws/tests/naming.js]: Should it really handle `undefined` or throw an error instead? (same for the `stage` etc.)

File: lib/plugins/aws/tests/naming.js [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+            serverless.service.service,
+            options.stage,
+            undefined,
+            'lambdaRole',
+          ],
+        ],
+      });
+    });
+  });
+  it('#getLogicalRoleName()', () => {
```
Comment (pmuens) [inline lib/plugins/aws/tests/naming.js]: Why is there no wrapping in a `describe` block?

File: lib/plugins/aws/tests/naming.js [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+          '-',
+          [
+            undefined,
+            serverless.service.service,
+            'lambda',
+          ],
+        ],
+      });
+    });
+    it('handles undefined service name', () => {
```
Comment (pmuens) [inline lib/plugins/aws/tests/naming.js]: Should it handle the `undefined` case?

File: lib/plugins/create/create.js [hunk 12]
```diff
@@ -85,7 +85,7 @@ class Create {
 
     // rename the service if the user has provided a path via options
     if (servicePath || serviceName) {
-      const newServiceName = serviceName || servicePath.split(path.sep).pop();
+      const newServiceName = serviceName || servicePath.split(path.sep).pop(); // NAME_HERE
```
Comment (pmuens) [inline lib/plugins/create/create.js]: What does this mean?

File: lib/plugins/package/lib/packageService.js [hunk 13]
```diff
@@ -22,11 +22,11 @@ module.exports = {
   },
 
   getServiceArtifactName() {
-    return `${this.serverless.service.service}.zip`;
+    return `${this.serverless.service.service}.zip`; // NAME_HERE
```
Comment (pmuens) [inline lib/plugins/package/lib/packageService.js]: What does this comment mean?

File: lib/plugins/package/lib/packageService.js [hunk 14]
```diff
@@ -22,11 +22,11 @@ module.exports = {
...
   },
 
   getFunctionArtifactName(functionObject) {
-    return `${functionObject.name}.zip`;
+    return `${functionObject.name}.zip`; // NAME_HERE
```
Comment (pmuens) [inline lib/plugins/package/lib/packageService.js]: What does this comment mean?

File: lib/plugins/aws/lib/naming.js [hunk 15]
```diff
@@ -0,0 +1,225 @@
+'use strict';
```
Comment (pmuens) [inline lib/plugins/aws/lib/naming.js]: Should this file live in `lib`? or should we move it to a `utils` directory as it's kind of a utility functionality...

**Comment by erikerikson** [comment]
Thank you for the feedback and sorry @pmuens, I marked this as not ready for review.  I will be focusing very strongly on this PR starting Monday, if not sooner.

File: lib/plugins/package/lib/packageService.js (hunk 14 above)
Comment (erikerikson) [inline lib/plugins/package/lib/packageService.js]: The // NAME_HERE comments indicate a name, possible name, or possible naming convention and should be removed prior to merge.

**Comment by pmuens** [comment]
Thanks for the feedback @erikerikson! No worries.

Just took a look into this as we're pushing it hard so that it can be released next week (as it's very valuable). Thanks for all the hard work and all your contributions! 👑

**Comment by eahefnawy** [comment]
@erikerikson thanks for your all effort here 😊 . I've just merged your changes after fixing all the conflicts in #2679 ... So closing this one 🙌
